money-akka/src/main/scala/com/comcast/money/akka/SpanContextWithStack.scala

```diff
@@ -79,28 +81,10 @@ class SpanContextWithStack() extends SpanContext {
    * @param span the [[Span]] to be prepended
    */
 
-  override def push(span: Span): Unit = stack = span :: stack
-
-  /**
-   * CAUTION THIS METHOD WILL NOT TRACK THE SPAN
-   *
-   * Returns the last element inserted and removes it
-   *
-   * [[Some]] when Stack has at least one element
-   *
-   * [[None]] when Stack is empty
-   *
-   * @return Option[Span]
-   */
-
-  override def pop(): Option[Span] =
-    stack.headOption match {
-      case someSpan: Some[Span] =>
-        stack = stack.tail
-        someSpan
-
-      case None => None
-    }
+  override def push(span: Span): Scope = {
+    stack = span :: stack
+    () => stack = stack.drop(1)
```

Comment:
can you explain what Scope is here?  Looks like we are mutating the thread local stack by removing head, I surmise this is `Unit` as well

Reply:
`Scope` is a functional interface equivalent to `() => ()`.  The `push` function adds the span to the head of the list and then returns the function to pop the head from that list.

Reply:
@ptravers I'm less familiar with the span state that the money-akka project needs to track, do you mind checking this out to make sure that it won't mess anything up?

Reply:
I don't think the change would necessarily be correct. The end user could call these out of order and get the wrong span,

Push Span A get Scope A
Push Span B get Scope B
Call Scope A get Span B

Reply:
Yes, that was already a problem if you called `push`/`pop` out of order.  It's still a problem in OTel if you call `push`/`close` out of order.  At most `Context` logs that you're reattaching the wrong context, but it allows it to happen.

I could follow a similar approach as `Context` and log (or throw?) if you attempt to close a scope out of order by comparing the top span before popping.  Does that alleviate your concern?

Reply:
This is the comment from grpc Context if you attempt to pop the incorrect scope:

```java
    if (current() != toDetach) {
      // Log a severe message instead of throwing an exception as the context to attach is assumed
      // to be the correct one and the unbalanced state represents a coding mistake in a lower
      // layer in the stack that cannot be recovered from here.
      log.log(Level.SEVERE, "Context was not attached when detaching",
          new Throwable().fillInStackTrace());
    }
```

Reply:
How does this approach feel?  It's similar to Context where it will log if the current state and expected state don't match indicating a stack imbalance.  If there is an unbalance it will still restore the previous state.

```scala
  override def push(span: Span): Scope = {
    // capture the previous stack of spans
    val previousStack = stack
    // push the new span to the head of the stack
    stack = span :: previousStack

    () => stack match {
      // happy path, the remaining stack of spans matches the captured stack of spans
      case _ :: rest if rest == previousStack =>
        stack = rest
      // unbalanced state, log the expected/actual head span and restore the stack to the captured stack of spans
      case head :: _ =>
        logSpanMismatch(span, head)
        stack = previousStack
      // unbalanced state, log the expected head span and leave the stack of spans empty
      case Nil =>
        logSpanMismatch(span, null)
    }
  }
```

Reply:
Apologies for a slow reply. That works for me